GitUI/UserControls/NativeTreeView.cs

```diff
@@ -8,4 +12,100 @@ protected override void CreateHandle()
             NativeMethods.SetWindowTheme(Handle, "explorer", null);
         }
     }
+
+    // Add double-click related behaviour to NativeTreeView
+    public class NativeTreeViewDoubleClickDecorator
+    {
+        private NativeTreeView _treeView;
+
+        private DateTime _lastMouseDown = DateTime.MinValue;
+        private int _mouseClicks = 0;
+
+        private DateTime _lastDoubleClick = DateTime.MinValue;
+
+        // Invoked just before an inner tree node is expanded/collapsed. Set CancelEventHandler.Cancel to true
+        // to cancel the expand/collapse of the current node.
+        public event CancelEventHandler BeforeDoubleClickExpandCollapse;
```

Comment:
I agree with you that the having "DoubleClick" in the event name is redundant within the context of this class name; however, this is also a suggestion for the name of the handler in the client class. For instance, in RepoObjectsTree, the function is named BeforeDoubleClickExpandCollapse, which is better than naming it BeforeExpandCollapse within _that_ context, since we want to know that this function gets called for double-click expand/collapse. So either we leave this name as is to get the suggested name (and VS will use that to auto-generate part of the function name), or we go with your idea and hope clients of the code come up with the right name.

Reply:
Fair enough, let's keep it.